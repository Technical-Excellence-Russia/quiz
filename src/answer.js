import React from "react"
import PropTypes from "prop-types"

export default function Answer(props) {

	const className = `row p-2 m-2 border rounded ${props.isOpen ? (props.answer.correct ? "border-success" : "border-danger") : "border-none"}`

	return (
		<div className={className}>
			<div className="col text-left form-check">
				<input id={props.index} type="checkbox" className="form-check-input"/>
				<label htmlFor={props.index} className="form-check-label">{props.answer.text}</label>
			</div>
		</div>
	)
}

Answer.propTypes = {
	index: PropTypes.string.isRequired,
	answer: PropTypes.object.isRequired,
	isOpen: PropTypes.bool.isRequired
}