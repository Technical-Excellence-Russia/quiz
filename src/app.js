import React, {useState} from "react"
import PropTypes from "prop-types"
import Answer from "./answer"

export default function App(props) {

	const total = 5
	const nextQuestion = () => props.questionFactory.getQuestion()

	const initialState = {
		questionIndex: 1,
		answerIndex: 0,
		isAnswerOpen: false,
		isResultOpen: false,
		question: nextQuestion()
	}

	const [state, setState] = useState(initialState)

	const isAnswersCorrect = () => {
		return state.question.answers.filter((answer) => {
			return answer.correct !== (answer.checked || false)
		}).length === 0
	}

	const putCheck = (index, value) => {
		const answers = state.question.answers
		answers[index].checked = value
		setState(
			{
				...state,
				question: {
					...state.question,
					answers: answers
				}
			}
		)
	}

	const showAnswer = () => setState({
			...state,
			isAnswerOpen: true
		}
	)

	const nextAnswer = () => {
		state.answerIndex = state.answerIndex + (isAnswersCorrect() ? 1 : 0)
		if (state.questionIndex === total) {
			setState({
					...state,
					isResultOpen: true
				}
			)
		} else {
			setState({
					...state,
					questionIndex: state.questionIndex + 1,
					isAnswerOpen: false,
					question: nextQuestion()
				}
			)
		}
	}

	const closeResult = () => setState(initialState)

	const goToGroup = () => window.open("https://t.me/technicalexcellenceru", "blank")

	const answers = state.question.answers.map((answer, index) => (
			<Answer key={index.toString()} index={index.toString()} answer={answer}
					isOpen={state.isAnswerOpen}
					onChange={(value) => putCheck(index, value)}/>
		)
	)

	const answerTag = (
		<>
			<div className={"row"}>
				<div className="col m-2 p-2 text-left border" dangerouslySetInnerHTML={{__html: state.question.description}}>

				</div>
			</div>
		</>
	)

	const progress = () => (state.questionIndex - 1) * 100 / total

	return (
		<>
			<div className={"row " + (state.isResultOpen ? "" : "d-none")}>
				<div className="col m-2 p-2 border-top border-light">
					Ты ответил правильно на {state.answerIndex} из {total}. <br/>
					А ещё у нас есть группа практиков;)
				</div>
			</div>

			<div className={"row " + (state.isResultOpen ? "d-none" : "")}>
				<div className="col m-2 p-2 border-top border-light">
					Вопрос {state.questionIndex}/{total}, правильно: {state.answerIndex}
					<div className="progress">
						<div className="progress-bar" role="progressbar" style={{width: `${progress()}%`}}
							 aria-valuenow={progress()}
							 aria-valuemin="0"
							 aria-valuemax="100">
						</div>
					</div>
				</div>
			</div>
			<div className={"row " + (state.isResultOpen ? "d-none" : "")}>
				<div className="col m-2 p-2 border-top border-light text-left text-light fw-bold">{state.question.text}</div>
			</div>
			{state.isResultOpen ? "" : answers}
			{state.isAnswerOpen ? answerTag : ""}
			<div className={"row " + (state.isResultOpen ? "d-none" : "")}>
				<div className="col m-2 p-2 border-top border-light text-center">
					<button className={`btn me-2 ${state.isAnswerOpen ? "btn-secondary" : "btn-primary"}`} type="button" onClick={() => showAnswer()}>
						Проверить
					</button>
					<button className={`btn ms-2 ${state.isAnswerOpen ? "btn-primary" : "btn-secondary"}`} type="button" onClick={() => nextAnswer()}>
						Следующий
					</button>
				</div>
			</div>
			<div className={"row " + (state.isResultOpen ? "" : "d-none")}>
				<div className="col m-2 p-2 border-top border-light text-center">
					<button className="btn me-2 btn-primary" type="button" onClick={() => goToGroup()}>
						В группу
					</button>
					<button className="btn ms-2 btn-secondary" type="button" onClick={() => closeResult()}>
						Ещё раз
					</button>
				</div>
			</div>
		</>
	)
}

App.propTypes = {
	questionFactory: PropTypes.object.isRequired
}