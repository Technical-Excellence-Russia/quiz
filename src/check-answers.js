export default function CheckAnswers(answers) {
	return answers.reduce((answer, acc) => acc &= answer.correct && (answer.checked || false))
}