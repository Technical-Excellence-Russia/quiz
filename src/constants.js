export const INIT = "INIT"
export const GO_TO_GROUP = "GO_TO_GROUP"
export const NEXT_QUESTION = "NEXT_QUESTION"
export const SET_CHECK = "SET_CHECK"
export const TOGGLE_ANSWER = "SHOW_ANSWER"