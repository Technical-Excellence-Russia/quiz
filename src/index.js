import "bootstrap"
import "./index.scss"

import React from "react"
import ReactDOM from "react-dom"

import QuestionFactory from "../lib/question"
import ReducerFactory from "./reducer-factory"
import App from "./app"

const currentYear = new Date().getFullYear()
const reducer = ReducerFactory(window, new QuestionFactory())

ReactDOM.render(
	<div className="container">
		<div className="row">
			<div className="co mt-1 text-center">
				<h2 className="text-center">Technical Excellence Quiz</h2>
			</div>
		</div>
		<App Reducer={reducer}/>
		<div className="row m-3">
			<div className="col text-end">
				<p className="copyright">&copy; {currentYear} Technical Excellence Ru</p>
			</div>
			<div className="col-sm-1">&nbsp;</div>
		</div>
	</div>
	, document.getElementById("app"))