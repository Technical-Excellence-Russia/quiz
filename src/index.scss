@import "~bootstrap/scss/bootstrap.scss";

$font-color: #d3d3d3;
$border-color: #939393;
$background-color: #2b2b2b;

html, body, {
  color: $font-color;
  background: $background-color;
  font-family: "Lucida Console", monospace;
}

code {
  color: $font-color;
}

table th, td {
  border: 1px solid $border-color;
  padding: 1px 1px 1px 1px;
}

.border-none {
  border-color: $background-color !important;
}

.copyright {
  font-size: 8pt;
}

@media (min-width: 240px) {
  font-size: 12px;
}

@media (min-width: 400px) {
  font-size: 14px;
}

@media (min-width: 768px) {
  font-size: 16px;
}

@media (min-width: 1000px) {
  font-size: 18px;
}

