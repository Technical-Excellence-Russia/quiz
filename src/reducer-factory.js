import {GO_TO_GROUP, INIT, NEXT_QUESTION, SET_CHECK, TOGGLE_ANSWER} from "./constants"
import CheckAnswers from "./check-answers"

export default function ReducerFactory(window, questionFactory) {
	const nextQuestion = () => questionFactory.getQuestion()

	const initialState = {
		questionIndex: 1,
		answerIndex: 0,
		totalQuestions: 5,
		isAnswerOpen: false,
		isResultOpen: false,
		question: nextQuestion()
	}

	return function Reducer(state, action) {
		switch (action.type) {
		case INIT: {
			return initialState
		}

		case SET_CHECK: {
			const answers = state.question.answers
			answers[action.index].checked = action.value
			return {
				...state,
				question: {
					...state.question,
					answers: answers
				}
			}
		}

		case TOGGLE_ANSWER: {
			return {
				...state,
				isAnswerOpen: !state.isAnswerOpen
			}
		}

		case NEXT_QUESTION: {
			state.answerIndex = state.answerIndex + (CheckAnswers(state.question.answers) ? 1 : 0)
			if (state.questionIndex === state.totalQuestions) {
				return {
					...state,
					isResultOpen: true
				}
			} else {
				return {
					...state,
					questionIndex: state.questionIndex + 1,
					isAnswerOpen: false,
					question: nextQuestion()
				}
			}
		}

		case GO_TO_GROUP: {
			window.open("https://t.me/technicalexcellenceru", "blank")
			return state
		}

		default: {
			return state
		}
		}
	}

}