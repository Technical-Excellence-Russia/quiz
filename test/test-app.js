import React from "react"
import renderer from "react-test-renderer"

import App from "../src/app"
import {INIT} from "../src/constants"

jest.mock("../src/components/question", () => () => "Question")
jest.mock("../src/components/result", () => () => "Result")

describe("App should", () => {

	let component = null

	test("equal to snapshot when result is closed", () => {
		const reducer = (_, action) => {
			expect(action).toEqual({type: INIT})
			return {isResultOpen: false}
		}
		component = renderer.create(<App Reducer={reducer}/>)
		expect(component.toJSON()).toMatchSnapshot()
		expect.assertions(2)
	})

	test("equal to snapshot when result is open", () => {
		const reducer = (_, action) => {
			expect(action).toEqual({type: INIT})
			return {isResultOpen: true}
		}
		component = renderer.create(<App Reducer={reducer}/>)
		expect(component.toJSON()).toMatchSnapshot()
		expect.assertions(2)
	})


	afterAll(() => component.unmount())
})