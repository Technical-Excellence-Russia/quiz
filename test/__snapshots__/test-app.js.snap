// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should equal to snapshot when result is closed 1`] = `"Question"`;

exports[`App should equal to snapshot when result is open 1`] = `"Result"`;
